go/libraries/doltcore/type_conversion.go

```diff
@@ -59,66 +60,82 @@ func GetUnderlyingError(err error) error {
-		types.BoolKind:   nil,
-		types.NullKind:   convToNullFunc},
+		types.StringKind:          convUUIDToString,
+		types.UUIDKind:            identityConvFunc,
+		types.UintKind:            nil,
+		types.IntKind:             nil,
+		types.FloatKind:           nil,
+		types.BoolKind:            nil,
+		types.UnderlyingArrayKind: nil,
+		types.NullKind:            convToNullFunc},
 	types.UintKind: {
-		types.StringKind: convUintToString,
-		types.UUIDKind:   nil,
-		types.UintKind:   identityConvFunc,
-		types.IntKind:    convUintToInt,
-		types.FloatKind:  convUintToFloat,
-		types.BoolKind:   convUintToBool,
-		types.NullKind:   convToNullFunc},
+		types.StringKind:          convUintToString,
+		types.UUIDKind:            nil,
+		types.UintKind:            identityConvFunc,
+		types.IntKind:             convUintToInt,
+		types.FloatKind:           convUintToFloat,
+		types.BoolKind:            convUintToBool,
+		types.UnderlyingArrayKind: convUintToUnderlyingArray,
+		types.NullKind:            convToNullFunc},
 	types.IntKind: {
-		types.StringKind: convIntToString,
-		types.UUIDKind:   nil,
-		types.UintKind:   convIntToUint,
-		types.IntKind:    identityConvFunc,
-		types.FloatKind:  convIntToFloat,
-		types.BoolKind:   convIntToBool,
-		types.NullKind:   convToNullFunc},
+		types.StringKind:          convIntToString,
+		types.UUIDKind:            nil,
+		types.UintKind:            convIntToUint,
+		types.IntKind:             identityConvFunc,
+		types.FloatKind:           convIntToFloat,
+		types.BoolKind:            convIntToBool,
+		types.UnderlyingArrayKind: convIntToUnderlyingArray,
+		types.NullKind:            convToNullFunc},
 	types.FloatKind: {
-		types.StringKind: convFloatToString,
-		types.UUIDKind:   nil,
-		types.UintKind:   convFloatToUint,
-		types.IntKind:    convFloatToInt,
-		types.FloatKind:  identityConvFunc,
-		types.BoolKind:   convFloatToBool,
-		types.NullKind:   convToNullFunc},
+		types.StringKind:          convFloatToString,
+		types.UUIDKind:            nil,
+		types.UintKind:            convFloatToUint,
+		types.IntKind:             convFloatToInt,
+		types.FloatKind:           identityConvFunc,
+		types.BoolKind:            convFloatToBool,
+		types.UnderlyingArrayKind: convFloatToUnderlyingArray,
+		types.NullKind:            convToNullFunc},
 	types.BoolKind: {
-		types.StringKind: convBoolToString,
-		types.UUIDKind:   nil,
-		types.UintKind:   convBoolToUint,
-		types.IntKind:    convBoolToInt,
-		types.FloatKind:  convBoolToFloat,
-		types.BoolKind:   identityConvFunc,
-		types.NullKind:   convToNullFunc},
+		types.StringKind:          convBoolToString,
+		types.UUIDKind:            nil,
+		types.UintKind:            convBoolToUint,
+		types.IntKind:             convBoolToInt,
+		types.FloatKind:           convBoolToFloat,
+		types.BoolKind:            identityConvFunc,
+		types.UnderlyingArrayKind: convBoolToUnderlyingArray,
+		types.NullKind:            convToNullFunc},
+	types.UnderlyingArrayKind: {
```

Comment:
I don't think these conversions make sense. The array type is only the storage details for some concrete type. You don't "convert" between a binary format and a high-level type, you "marshall" and "unmarshall". To put it another way, it makes sense to convert between a string and a timestamp (which is built on top of a byte array), but not between a string the underlying byte array.

Reply:
It is marshalling between types. Further down in the same file you'll see the implementation of each of these.